tls/s2n_handshake_io.c

```diff
@@ -1003,20 +1003,29 @@ static int s2n_handshake_read_io(struct s2n_connection *conn)
 
 static int s2n_handle_retry_state(struct s2n_connection *conn)
 {
-    /* If we were blocked reading a record, then the handler is waiting on external data.
-     * The handler will know how to continue, so we should call the handler right away.
-     * We aren't going to read more handshake data yet because the current message has
-     * not finished processing. */
-    s2n_errno = S2N_ERR_T_OK;
+    /* If we were blocked reading or writing a record, then the handler is waiting on
+     * external data. The handler will know how to continue, so we should call the
+     * handler right away. We aren't going to read more handshake data yet or proceed
+     * to the next handler because the current message has not finished processing. */
+    s2n_errno = S2N_ERR_OK;
     const int r = ACTIVE_STATE(conn).handler[conn->mode] (conn);
 
-    if (r < 0) {
+    if (r < 0 && S2N_ERROR_IS_BLOCKING(s2n_errno)) {
         /* If the handler is still waiting for data, return control to the caller. */
-        if (S2N_ERROR_IS_BLOCKING(s2n_errno)) {
-            S2N_ERROR_PRESERVE_ERRNO();
-        }
+        S2N_ERROR_PRESERVE_ERRNO();
+    }
+
+    const char this = conn->mode == S2N_CLIENT ? 'C' : 'S';
```

Comment:
Could we do a clearer variable name over `this`? Something like `connection_mode` or `local_mode`?

Reply:
Sure, just followed previous convention in s2n :D I'll change it to `conmnection_mode`

Reply:
I'm seeing that now... That's unfortunate :(